# 1. Постановка задачи

На строительство магистрали периодически доставляются материалы, которые со станции железной  
дороги вначале поступают на $3$ промежуточных склада, а затем непосредственно к $5$ объектам магистрали.  
Два первых склада $А1$, $А2$ имеют ограниченную емкость $50[т]$, и поступившие на них грузы расходуются  
полностью. Cклад $А3$ с неограниченной емкостью допускает резервирование грузов от периода к периоду.  
Транспортные расходы (в условных денежных единицах — $[ДЕ]$) на перевозку одной тонны груза со станции  
на промежуточные склады составляют $7$, $4$, $4[ДЕ]$, а со складов к объектам $В1–В5$ представлены  
в табл. $1$; там же приведены потребности объектов (в тоннах) в течение двух периодов.

<table style="display: inline-block">
    <caption>Таблица $1$: Транспортные расходы</caption>
    <tr>
        <th></th>
        <th>$B1$</th>
        <th>$B2$</th>
        <th>$B3$</th>
        <th>$B4$</th>
        <th>$B5$</th>
    </tr>
    <tr>
        <td>$A1$</td>
        <td>$7$</td>
        <td>$2$</td>
        <td>$13$</td>
        <td>$11$</td>
        <td>$11$</td>
    </tr>
    <tr>
        <td>$A2$</td>
        <td>$11$</td>
        <td>$2$</td>
        <td>$1$</td>
        <td>$9$</td>
        <td>$6$</td>
    </tr>
    <tr>
        <td>$A3$</td>
        <td>$8$</td>
        <td>$11$</td>
        <td>$1$</td>
        <td>$13$</td>
        <td>$13$</td>
    </tr>
    <tr>
        <td>$1й\ период$</td>
        <td>$37$</td>
        <td>$26$</td>
        <td>$31$</td>
        <td>$34$</td>
        <td>$16$</td>
    </tr>
    <tr>
        <td>$2й\ период$</td>
        <td>$57$</td>
        <td>$13$</td>
        <td>$2$</td>
        <td>$19$</td>
        <td>$47$</td>
    </tr>
</table>

Рассматривается работа системы в течение двух периодов при условии доставки на станцию по $150[т]$  
груза в каждый из периодов. Требуется определить план перевозок, обеспечивающий минимум стоимости  
доставки грузов для двух периодов. Провести анализ чувствительности плана к изменению цен перевозок  
из первого промежуточного склада.

# 2. Формализация задачи

* Вектор $\overline{c}_{18 \times 1}$ обозначает стоимость перевозок (в $[ДЕ]$),  
где первые $3$ компоненты - стоимость перевозки в промежуточные склады,  
последующие - стоимости перевозок по строкам таблицы, слева направо;
* Вектор $\overline{x}_{18 \times 1}$ обозначает количество (в $[т]$) товара, которое необходимо распределить.  
Компоненты соответствуют компонентам из вектора $\overline{c}_{18 \times 1}$ по смыслу "откуда-куда";  
* Матрица $G_{21 \times 18}$ обозначает коэффициенты при компонентах $\overline{x}$ в левой части ограничений на неравенства ($\leq$);
* Вектор $\overline{h}_{21 \times 1}$ обозначает коэффициенты в правой части ограничений на неравенства ($\leq$);
* Матрица $A_{8 \times 18}$ обозначает коэффициенты при компонентах $\overline{x}$ в левой части ограничений на равенства ($=$);
* Вектор $\overline{b}_{8 \times 1}$ обозначает коэффициенты в правой части ограничений на равенства ($=$).

Перейдём к системам для периодов:



$period_1\begin{cases}f(\overline{x}^{(1)})=7x_1^{(1)} + 4x_2^{(1)} + 4x_3^{(1)} \\
+7x_4^{(1)} + 2x_5^{(1)} + 13x_6^{(1)} + 11x_7^{(1)} + + 11x_8^{(1)} \\
+11x_9^{(1)} + 2x_{10}^{(1)} + 1x_{11}^{(1)} + 9x_{12}^{(1)} + 6x_{13}^{(1)} \\
+8x_{14}^{(1)} + 11x_{15}^{(1)}+ 1x_{16}^{(1)} + 13x_{17}^{(1)} + 13x_{18}^{(1)} \rightarrow \min \\
x_1^{(1)} + x_2^{(1)} + x_3^{(1)} = 150 \\
7x_4^{(1)}  + 11x_9^{(1)}   + 8x_{14}^{(1)}  = 37 \\
2x_5^{(1)}  + 2x_{10}^{(1)} + 11x_{15}^{(1)} = 26 \\
13x_6^{(1)} + 1x_{11}^{(1)} + 1x_{16}^{(1)}  = 31 \\
11x_7^{(1)} + 9x_{12}^{(1)} + 13x_{17}^{(1)} = 34 \\
11x_8^{(1)} + 6x_{13}^{(1)} + 13x_{18}^{(1)} = 16 \\
x_1^{(1)} = \sum\limits_{i=4}^8x_i^{(1)}    \\
x_2^{(1)} = \sum\limits_{i=9}^{13}x_i^{(1)} \\
\sum\limits_{i=4}^{18}x_i^{(2)} \leq 150 + r^{(1)} \\
x_1^{(1)},x_2^{(1)} \leq 50 \\
x_i^{(1)} \geq 0, i=\overline{0,18} \\
\end{cases}period_2\begin{cases}
f(\overline{x}^{(2)})=7x_1^{(2)} + 4x_2^{(2)} + 4x_3^{(2)} \\
+7x_4^{(2)}    + 2x_5^{(2)}     + 13x_6^{(2)}   + 11x_7^{(2)}    + 11x_8^{(2)} \\
+11x_9^{(2)}   + 2x_{10}^{(2)}  + 1x_{11}^{(2)} + 9x_{12}^{(2)}  + 6x_{13}^{(2)} \\
+8x_{14}^{(2)} + 11x_{15}^{(2)} + 1x_{16}^{(2)} + 13x_{17}^{(2)} + 13x_{18}^{(2)} \rightarrow \min \\
x_1^{(2)} + x_2^{(2)} + x_3^{(2)} = 150 \\
7x_4^{(2)}  + 11x_9^{(2)}   + 8x_{14}^{(2)}  = 57 \\
2x_5^{(2)}  + 2x_{10}^{(2)} + 11x_{15}^{(2)} = 13 \\
13x_6^{(2)} + 1x_{11}^{(2)} + 1x_{16}^{(2)}  = 2 \\
11x_7^{(2)} + 9x_{12}^{(2)} + 13x_{17}^{(2)} = 19 \\
11x_8^{(2)} + 6x_{13}^{(2)} + 13x_{18}^{(2)} = 47 \\
x_1^{(2)} = \sum\limits_{i=4}^8x_i^{(2)} \\
x_2^{(2)} = \sum\limits_{i=9}^{13}x_i^{(2)} \\
\sum\limits_{i=4}^{18}x_i^{(2)} \leq 150 + r^{(2)} \\
x_1^{(2)},x_2^{(2)} \leq 50 \\
x_i^{(2)} \geq 0, i=\overline{0,18} \\
\end{cases}$

$\begin{cases}
\overline{c}^T\overline{x}^{(j)}\rightarrow\min \\
G\overline{x}^{(j)}\leq h^{(j)} \\
A\overline{x}^{(j)} = b^{(j)} \\
\end{cases},j=\overline{1,2}$

# 3. Описание задачи

Для решения задачи была выбрана библиотека `CVXOPT`.

В данной библиотеке присутствует функция `lp` для решения задач ЛП.  
Ей передаём:
* транспортные расходы ($\overline{c}_{18 \times 1}$); 
* ограничения на неравенства ($G_{21 \times 18}$ и $\overline{h}_{21 \times 1}$);
* а также ограничения на равенства ($A_{8 \times 18}$ и $\overline{b}_{8 \times 1}$).

Вызываем эту функцию с параметром `solver = 'glpk'`, т.е. будем использовать симплекс-метод.  
Сохраняем значение остатка на промежуточном складе $A3$, а также общую стоимость перевозок.  
На будущее сохраним поле `['z']` для анализа чувствительности.

Далее распишем матрицы, отображающие начальные условия для `CVXOPT`.

$
\overline{c}_{18 \times 1} =
\begin{pmatrix}
7\\
4\\
4\\
7\\
2\\
13\\
11\\
11\\
11\\
2\\
1\\
9\\
6\\
8\\
11\\
1\\
13\\
13
\end{pmatrix}\quad
\overline{x}_{18 \times 1} =
\begin{pmatrix}
x_1\\
x_2\\
x_3\\
x_4\\
x_5\\
x_6\\
x_7\\
x_8\\
x_9\\
x_{10}\\
x_{11}\\
x_{12}\\
x_{13}\\
x_{14}\\
x_{15}\\
x_{16}\\
x_{17}\\
x_{18}
\end{pmatrix}\quad
\overline{h}^0_{21 \times 1} =
\begin{pmatrix}
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
0\\
50\\
50\\
150\\
\end{pmatrix}\quad
\overline{b}^{(1)}_{8 \times 1} =
\begin{pmatrix}
150\\
37\\
26\\
31\\
34\\
16\\
0\\
0
\end{pmatrix}\quad
\overline{b}^{(2)}_{8 \times 1} =
\begin{pmatrix}
150\\
57\\
13\\
2\\
19\\
47\\
0\\
0
\end{pmatrix}
\\
G_{21 \times 18} =
\begin{pmatrix}
-1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & -1 \\
1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 1 & 1 & 1 & 1 & 1 & 1 & 1 & 1 & 1 & 1 & 1 & 1 & 1 & 1 & 1 \\
\end{pmatrix}
\\
A_{8 \times 18} =
\begin{pmatrix}
1 & 1 & 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 \\
-1 & 0 & 0 & 1 & 1 & 1 & 1 & 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
0 & -1 & 0 & 0 & 0 & 0 & 0 & 0 & 1 & 1 & 1 & 1 & 1 & 0 & 0 & 0 & 0 & 0 \\
\end{pmatrix}                                                                 \\\\
\begin{cases}
    f(\overline{x}^{(j)}) = (\overline{c}^T, \overline{x}^{(j)})
    \rightarrow \min                                                           \\\\
    G\overline{x}^{(j)} \leq \overline{h}^{(j)}                                           \\\\
    A\overline{x}^{(j)} = \overline{b}^{(j)}                                              \\\\
    \overline{h}^{(j)} = \overline{h}^0, \overline{h}^{(j)}_{21} = 150 + r^{(j)}
\end{cases}
\qquad j = [1,2],
\ r^{(1)} = 0,\ r^{(2)} = \overline{h}^{(1)}_{21} - \sum \limits_{n=4}^{18} \overline{x}^{(1)}_n
$

# 4. Решение задачи

In [1]:
import numpy as np
from cvxopt   import matrix, solvers
from tabulate import tabulate

def solve(c, G, h, A, b, r, p, d, i):
    so     = solvers.lp(c, G, h, A, b, solver = 'glpk')
    d[i]   = (so['z'][3:8])
    st     = so['status']
    if st == 'optimal':
        A     = so['x'][0:3]
        B     = [x for j in range(0,5) for x in so['x'][3 + j:15 + j:5]]
        B1    = B[0:3]
        B2    = B[3:6]
        B3    = B[6:9]
        B4    = B[9:12]
        B5    = B[12:15]
        B1S   = sum(B1)
        B2S   = sum(B2)
        B3S   = sum(B3)
        B4S   = sum(B4)
        B5S   = sum(B5)
        A3    = so['x'][13:18]
        O     = np.append(np.hstack(A), sum(A))
        A[2]  += r[i]
        T     = np.append(np.hstack(A), sum(A))
        R     = np.array([sum(so['x'][3:8]), sum(so['x'][8:13]), sum(A3)])
        t     = tabulate(
            np.array([
                np.array([ 'A1 ',  'A2 ',  'A3 ',  'Σ ']),
                O,
                T,
                np.append(B1, B1S),
                np.append(B2, B2S),
                np.append(B3, B3S),
                np.append(B4, B4S),
                np.append(B5, B5S),
                T - np.append(R, sum(R))
            ]).T,
            [f'P{i + 1} ',  'O ',  'T ',  'B1 ',  'B2 ',  'B3 ',  'B4 ',  'B5 ',  'R '],
            tablefmt = 'html'
        )
        display(t)
        p[i]   = (so['primal objective'])
        print(f'Objective:\t\t{p[i]}\n ')
        h[20] = b[0] + T[3] - (B1S + B2S + B3S + B4S + B5S)
        r[i + 1] = (h[20] - b[0])
    else:
        print(f'Cant\'t find optimal solution: {st}\n')

c = matrix([
    7., 4,  4,
    7,  2,  13, 11, 11,
    11, 2,  1,  9,  6,
    8,  11, 1,  13, 13
])
G = matrix([
    [-1.,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,-1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,0,-1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,0,0,-1,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,0,0,0,-1,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,-1,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,-1,0,0,0,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,0,-1,0,0,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,-1,0,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,-1,0,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,-1,0,0,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,-1,0,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,-1,0,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,-1,0,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,0,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,0,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,0],
    [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1],
    
    [1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
    [0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],
]).T
o = np.array([150.])
h = matrix([
    0.,  # -x_1  <= 0
    0,   # -x_2  <= 0
    0,   # -x_3  <= 0
    0,   # -x_4  <= 0
    0,   # -x_5  <= 0
    0,   # -x_6  <= 0
    0,   # -x_7  <= 0
    0,   # -x_8  <= 0
    0,   # -x_9  <= 0
    0,   # -x_10 <= 0
    0,   # -x_11 <= 0
    0,   # -x_12 <= 0
    0,   # -x_13 <= 0
    0,   # -x_14 <= 0
    0,   # -x_15 <= 0
    0,   # -x_16 <= 0
    0,   # -x_17 <= 0
    0,   # -x_18 <= 0
    
    50,   # x_1   <= 50
    50,   # x_2   <= 50
    o[0], # sum(_4,...,_18) <= 150 + r
])
A = matrix([
    [1.,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0], # _1 + _2 + _3 = 150
    [0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,0,0],  # _4 + _9 + _14 = 37 / 57
    [0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,0],  # _5 + _10 + _15 = 26 / 13
    [0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0],  # _6 + _11 + _16 = 31 / 2
    [0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0],  # _7 + _12 + _17 = 34 / 19
    [0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1],  # _8 + _13 + _18 = 16 / 47
    [-1,0,0,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0], # _1 = _4 + _5 + _6 + _7 + _8
    [0,-1,0,0,0,0,0,0,1,1,1,1,1,0,0,0,0,0], # _2 = _9 + _10 + _11 + _12 + _13
]).T
s = [[37, 26, 31, 34, 16],
     [57, 13, 2,  19, 47],]
e = [0, 0]
r = [0] * 3
p = [0] * 2
d = [0] * 2

for i in range(len(s)):
    b = matrix(np.concatenate((o, s[i], e)))
    solve(c, G, h, A, b, r, p, d, i)

print(f'Overall price:\t\t{sum(p)} ')

'<table>\n<thead>\n<tr><th>P1   </th><th style="text-align: right;">  O </th><th style="text-align: right;">  T </th><th style="text-align: right;">  B1 </th><th style="text-align: right;">  B2 </th><th style="text-align: right;">  B3 </th><th style="text-align: right;">  B4 </th><th style="text-align: right;">  B5 </th><th style="text-align: right;">  R </th></tr>\n</thead>\n<tbody>\n<tr><td>A1   </td><td style="text-align: right;">  26</td><td style="text-align: right;">  26</td><td style="text-align: right;">    0</td><td style="text-align: right;">   26</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   0</td></tr>\n<tr><td>A2   </td><td style="text-align: right;">  50</td><td style="text-align: right;">  50</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   34</td><td style="text-align: right;">   16</td><td style="text-align: right;">   0</td></tr>\n<tr><td>A3   </td><td style="text-align: right;">  74</td><td style="text-align: right;">  74</td><td style="text-align: right;">   37</td><td style="text-align: right;">    0</td><td style="text-align: right;">   31</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   6</td></tr>\n<tr><td>Σ    </td><td style="text-align: right;"> 150</td><td style="text-align: right;"> 150</td><td style="text-align: right;">   37</td><td style="text-align: right;">   26</td><td style="text-align: right;">   31</td><td style="text-align: right;">   34</td><td style="text-align: right;">   16</td><td style="text-align: right;">   6</td></tr>\n</tbody>\n</table>'

Objective:		1459.0
 


'<table>\n<thead>\n<tr><th>P2   </th><th style="text-align: right;">  O </th><th style="text-align: right;">  T </th><th style="text-align: right;">  B1 </th><th style="text-align: right;">  B2 </th><th style="text-align: right;">  B3 </th><th style="text-align: right;">  B4 </th><th style="text-align: right;">  B5 </th><th style="text-align: right;">  R </th></tr>\n</thead>\n<tbody>\n<tr><td>A1   </td><td style="text-align: right;">  13</td><td style="text-align: right;">  13</td><td style="text-align: right;">    0</td><td style="text-align: right;">   13</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   0</td></tr>\n<tr><td>A2   </td><td style="text-align: right;">  50</td><td style="text-align: right;">  50</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">    3</td><td style="text-align: right;">   47</td><td style="text-align: right;">   0</td></tr>\n<tr><td>A3   </td><td style="text-align: right;">  87</td><td style="text-align: right;">  93</td><td style="text-align: right;">   57</td><td style="text-align: right;">    0</td><td style="text-align: right;">    2</td><td style="text-align: right;">   16</td><td style="text-align: right;">    0</td><td style="text-align: right;">  18</td></tr>\n<tr><td>Σ    </td><td style="text-align: right;"> 150</td><td style="text-align: right;"> 156</td><td style="text-align: right;">   57</td><td style="text-align: right;">   13</td><td style="text-align: right;">    2</td><td style="text-align: right;">   19</td><td style="text-align: right;">   47</td><td style="text-align: right;">  18</td></tr>\n</tbody>\n</table>'

Objective:		1640.0
 
Overall price:		3099.0 


# 5. Интерпретация решения

В $1$ период необходимо перевезти из станции:
* $26[т]$ в $A1$;
* $50[т]$ в $A2$;
* $74[т]$ в $A3$.  

Далее, надо вывезти из:
* $A1$: $26[т]$ в $B2$; 
* $A2$: $34[т]$ в $B4$, $16[т]$ в $B5$;
* $A3$: $37[т]$ в $B1$, $31[т]$ в $B3$.  

Суммарная стоимость этих перевозок будет составлять $1459[ДЕ]$.  
При этом на $A3$ останется $6[т]$ товара.  

Во $2$ период необходимо вывезти из станции:
* $13[т]$ в $A1$;
* $50[т]$ в $A2$;
* $87[т]$ в $A3$
(с учётом $6[т]$, оставшихся с $1$ периода, всего на $A3$ будет $93[т]$).  

Далее, надо вывезти из:
* $A1$: $13[т]$ в $B2$; 
* $A2$: $3[т]$ в $B4$, $47[т]$ в $B5$; 
* $A3$: $57[т]$ в $B1$, $2[т]$ в $B3$, $16[т]$ в $B4$.  

Суммарная стоимость этих перевозок будет составлять $1640[ДЕ]$.  
При этом на $A3$ останется $18[т]$ товара.  

Суммарная стоимость перевозок за два периода будет составлять $3099[ДЕ]$.  
После осуществления перевозок на $A3$ останется $18[т]$ товара.

# 6. Исходные данные для анализа чувствительности
# и их интерпретация

С учётом изменения цен перевозок из первого промежуточного склада получаем следующее:

In [2]:
for i in range(len(d)):
    rcs = np.array([*np.array([ 'A1 ']),*np.array(d[i]).T[0]]).T,
    t = tabulate(rcs,
                [f'P{i + 1} ',  'B1 ',  'B2 ',  'B3 ',  'B4 ',  'B5 '],
                tablefmt = 'html'
        )
    print('Reduced costs:')
    display(t)

Reduced costs:


'<table>\n<thead>\n<tr><th>P1   </th><th style="text-align: right;">  B1 </th><th style="text-align: right;">  B2 </th><th style="text-align: right;">  B3 </th><th style="text-align: right;">  B4 </th><th style="text-align: right;">  B5 </th></tr>\n</thead>\n<tbody>\n<tr><td>A1   </td><td style="text-align: right;">    2</td><td style="text-align: right;">   -0</td><td style="text-align: right;">   15</td><td style="text-align: right;">    1</td><td style="text-align: right;">    4</td></tr>\n</tbody>\n</table>'

Reduced costs:


'<table>\n<thead>\n<tr><th>P2   </th><th style="text-align: right;">  B1 </th><th style="text-align: right;">  B2 </th><th style="text-align: right;">  B3 </th><th style="text-align: right;">  B4 </th><th style="text-align: right;">  B5 </th></tr>\n</thead>\n<tbody>\n<tr><td>A1   </td><td style="text-align: right;">    2</td><td style="text-align: right;">   -0</td><td style="text-align: right;">   15</td><td style="text-align: right;">    1</td><td style="text-align: right;">    4</td></tr>\n</tbody>\n</table>'

Таблица говорит вот о чём: при единичном изменении потребности в материалах для конкретного склада  
(отдельно для этого склада за раз, так и для нескольких) от промежуточного  
суммарная стоимость будет увеличиваться (или уменьшаться) на соответствующую в таблице величину.  
Попробуем теперь провести эксперимент для $1$ периода, и установить потребность в материалах для  
$B1$ от $A1$ большей, либо равной $1[т]$, т.к. ранее туда мы поставляли $0[т]$.  
Таблица говорит нам, что при этом суммарная стоимость должна вырасти на $2[ДЕ]$.

In [3]:
j = 0
b = matrix(np.concatenate((o, s[j], e)))
h[3] = -1
r = [0] * 3
p = [0] * 3
d = [0] * 3
solve(c, G, h, A, b, r, p, d, j)
h[3] = 0

'<table>\n<thead>\n<tr><th>P1   </th><th style="text-align: right;">  O </th><th style="text-align: right;">  T </th><th style="text-align: right;">  B1 </th><th style="text-align: right;">  B2 </th><th style="text-align: right;">  B3 </th><th style="text-align: right;">  B4 </th><th style="text-align: right;">  B5 </th><th style="text-align: right;">  R </th></tr>\n</thead>\n<tbody>\n<tr><td>A1   </td><td style="text-align: right;">  27</td><td style="text-align: right;">  27</td><td style="text-align: right;">    1</td><td style="text-align: right;">   26</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   0</td></tr>\n<tr><td>A2   </td><td style="text-align: right;">  50</td><td style="text-align: right;">  50</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   34</td><td style="text-align: right;">   16</td><td style="text-align: right;">   0</td></tr>\n<tr><td>A3   </td><td style="text-align: right;">  73</td><td style="text-align: right;">  73</td><td style="text-align: right;">   36</td><td style="text-align: right;">    0</td><td style="text-align: right;">   31</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   6</td></tr>\n<tr><td>Σ    </td><td style="text-align: right;"> 150</td><td style="text-align: right;"> 150</td><td style="text-align: right;">   37</td><td style="text-align: right;">   26</td><td style="text-align: right;">   31</td><td style="text-align: right;">   34</td><td style="text-align: right;">   16</td><td style="text-align: right;">   6</td></tr>\n</tbody>\n</table>'

Objective:		1461.0
 


Как видно, это действительно так. Теперь увеличим значение до $2[т]$.

In [4]:
j = 0
b = matrix(np.concatenate((o, s[j], e)))
h[3] = -2
r = [0] * 3
p = [0] * 3
d = [0] * 3
solve(c, G, h, A, b, r, p, d, j)
h[3] = 0

'<table>\n<thead>\n<tr><th>P1   </th><th style="text-align: right;">  O </th><th style="text-align: right;">  T </th><th style="text-align: right;">  B1 </th><th style="text-align: right;">  B2 </th><th style="text-align: right;">  B3 </th><th style="text-align: right;">  B4 </th><th style="text-align: right;">  B5 </th><th style="text-align: right;">  R </th></tr>\n</thead>\n<tbody>\n<tr><td>A1   </td><td style="text-align: right;">  28</td><td style="text-align: right;">  28</td><td style="text-align: right;">    2</td><td style="text-align: right;">   26</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   0</td></tr>\n<tr><td>A2   </td><td style="text-align: right;">  50</td><td style="text-align: right;">  50</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   34</td><td style="text-align: right;">   16</td><td style="text-align: right;">   0</td></tr>\n<tr><td>A3   </td><td style="text-align: right;">  72</td><td style="text-align: right;">  72</td><td style="text-align: right;">   35</td><td style="text-align: right;">    0</td><td style="text-align: right;">   31</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   6</td></tr>\n<tr><td>Σ    </td><td style="text-align: right;"> 150</td><td style="text-align: right;"> 150</td><td style="text-align: right;">   37</td><td style="text-align: right;">   26</td><td style="text-align: right;">   31</td><td style="text-align: right;">   34</td><td style="text-align: right;">   16</td><td style="text-align: right;">   6</td></tr>\n</tbody>\n</table>'

Objective:		1463.0
 


Правильно (не забываем, что изначальная стоимость была $1459[ДЕ]$). До $20[т]$.

In [5]:
j = 0
b = matrix(np.concatenate((o, s[j], e)))
h[3] = -20
r = [0] * 3
p = [0] * 3
d = [0] * 3
solve(c, G, h, A, b, r, p, d, j)
h[3] = 0

'<table>\n<thead>\n<tr><th>P1   </th><th style="text-align: right;">  O </th><th style="text-align: right;">  T </th><th style="text-align: right;">  B1 </th><th style="text-align: right;">  B2 </th><th style="text-align: right;">  B3 </th><th style="text-align: right;">  B4 </th><th style="text-align: right;">  B5 </th><th style="text-align: right;">  R </th></tr>\n</thead>\n<tbody>\n<tr><td>A1   </td><td style="text-align: right;">  46</td><td style="text-align: right;">  46</td><td style="text-align: right;">   20</td><td style="text-align: right;">   26</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   0</td></tr>\n<tr><td>A2   </td><td style="text-align: right;">  50</td><td style="text-align: right;">  50</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   34</td><td style="text-align: right;">   16</td><td style="text-align: right;">   0</td></tr>\n<tr><td>A3   </td><td style="text-align: right;">  54</td><td style="text-align: right;">  54</td><td style="text-align: right;">   17</td><td style="text-align: right;">    0</td><td style="text-align: right;">   31</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   6</td></tr>\n<tr><td>Σ    </td><td style="text-align: right;"> 150</td><td style="text-align: right;"> 150</td><td style="text-align: right;">   37</td><td style="text-align: right;">   26</td><td style="text-align: right;">   31</td><td style="text-align: right;">   34</td><td style="text-align: right;">   16</td><td style="text-align: right;">   6</td></tr>\n</tbody>\n</table>'

Objective:		1499.0
 


И снова таблица нас не подводит. Теперь попробуем сделать похожее действие для $B3$ и $2$ периода.  
Увеличим потребность в материале до минимум $2[т]$.  
Таблица говорит нам, что при данном изменении суммарная стоимость  
должна вырасти на $2\cdot15[ДЕ]=30[ДЕ]$ ($1640[ДЕ]\rightarrow1670[ДЕ]$).

In [6]:
j = 1
b = matrix(np.concatenate((o, s[j], e)))
h[5] = -2
r = [0] * 3
p = [0] * 3
d = [0] * 3
solve(c, G, h, A, b, r, p, d, j)
h[5] = 0

'<table>\n<thead>\n<tr><th>P2   </th><th style="text-align: right;">  O </th><th style="text-align: right;">  T </th><th style="text-align: right;">  B1 </th><th style="text-align: right;">  B2 </th><th style="text-align: right;">  B3 </th><th style="text-align: right;">  B4 </th><th style="text-align: right;">  B5 </th><th style="text-align: right;">  R </th></tr>\n</thead>\n<tbody>\n<tr><td>A1   </td><td style="text-align: right;">  15</td><td style="text-align: right;">  15</td><td style="text-align: right;">    0</td><td style="text-align: right;">   13</td><td style="text-align: right;">    2</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   0</td></tr>\n<tr><td>A2   </td><td style="text-align: right;">  50</td><td style="text-align: right;">  50</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">    3</td><td style="text-align: right;">   47</td><td style="text-align: right;">   0</td></tr>\n<tr><td>A3   </td><td style="text-align: right;">  85</td><td style="text-align: right;">  85</td><td style="text-align: right;">   57</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   16</td><td style="text-align: right;">    0</td><td style="text-align: right;">  12</td></tr>\n<tr><td>Σ    </td><td style="text-align: right;"> 150</td><td style="text-align: right;"> 150</td><td style="text-align: right;">   57</td><td style="text-align: right;">   13</td><td style="text-align: right;">    2</td><td style="text-align: right;">   19</td><td style="text-align: right;">   47</td><td style="text-align: right;">  12</td></tr>\n</tbody>\n</table>'

Objective:		1670.0
 


Увеличим потребность в материалах сразу в $B1$ и $B3$ для $1$ периода до $10[т]$.

In [7]:
j = 0
b = matrix(np.concatenate((o, s[j], e)))
h[3] = -10
h[5] = -10
r = [0] * 3
p = [0] * 3
d = [0] * 3
solve(c, G, h, A, b, r, p, d, j)
h[3] = 0
h[5] = 0

'<table>\n<thead>\n<tr><th>P1   </th><th style="text-align: right;">  O </th><th style="text-align: right;">  T </th><th style="text-align: right;">  B1 </th><th style="text-align: right;">  B2 </th><th style="text-align: right;">  B3 </th><th style="text-align: right;">  B4 </th><th style="text-align: right;">  B5 </th><th style="text-align: right;">  R </th></tr>\n</thead>\n<tbody>\n<tr><td>A1   </td><td style="text-align: right;">  46</td><td style="text-align: right;">  46</td><td style="text-align: right;">   10</td><td style="text-align: right;">   26</td><td style="text-align: right;">   10</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   0</td></tr>\n<tr><td>A2   </td><td style="text-align: right;">  50</td><td style="text-align: right;">  50</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   34</td><td style="text-align: right;">   16</td><td style="text-align: right;">   0</td></tr>\n<tr><td>A3   </td><td style="text-align: right;">  54</td><td style="text-align: right;">  54</td><td style="text-align: right;">   27</td><td style="text-align: right;">    0</td><td style="text-align: right;">   21</td><td style="text-align: right;">    0</td><td style="text-align: right;">    0</td><td style="text-align: right;">   6</td></tr>\n<tr><td>Σ    </td><td style="text-align: right;"> 150</td><td style="text-align: right;"> 150</td><td style="text-align: right;">   37</td><td style="text-align: right;">   26</td><td style="text-align: right;">   31</td><td style="text-align: right;">   34</td><td style="text-align: right;">   16</td><td style="text-align: right;">   6</td></tr>\n</tbody>\n</table>'

Objective:		1629.0
 


$1459[ДЕ]+10\cdot2[ДЕ]+10\cdot15[ДЕ]=1629[ДЕ]$. QED